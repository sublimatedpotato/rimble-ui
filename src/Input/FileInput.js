import React, { Component } from 'react';
import styled from 'styled-components';

import Button from '../Button';

const StyledFileInput = styled('input')`
  & {
    width: 0;
    height: 0;
    opacity: 0;
    position: absolute;
    overflow: hidden;
    z-index: -1;
  }
`;

const StyledFileLabel = styled('label')`
  & {
    display: inline-block;
    position: relative;
    cursor: pointer;
  }
`;

class FileInput extends Component {
  constructor(props) {
    super(props);
    this.fileInput = React.createRef();
    this.state = {
      buttonLabel: 'Choose file…',
      files: false,
    };
  }

  handleClick = e => {
    this.fileInput.current.click();
  };

  handleChange = e => {
    let files = this.fileInput.current.files;

    this.setState({
      files: files.length ? true : false,
    });

    if (files && files.length > 1) {
      this.setState({
        buttonLabel: files.length + ' files selected',
      });
    } else if (files && files.length) {
      this.setState({
        buttonLabel: files[0].name,
      });
    } else {
      this.setState({
        buttonLabel: 'Choose file…',
      });
    }
  };

  render() {
    const { variant, ...props } = this.props;

    if (variant === 'unstyled') {
      return <input type={'file'} {...props} />;
    }

    return (
      <StyledFileLabel>
        <StyledFileInput
          type={'file'}
          {...props}
          ref={this.fileInput}
          onChange={this.handleChange}
        />
        <Button
          icon={this.state.files ? 'CloudDone' : 'CloudUpload'}
          onClick={this.handleClick}
        >
          {this.state.buttonLabel}
        </Button>
      </StyledFileLabel>
    );
  }
}

FileInput.displayName = 'FileInput';

export default FileInput;
