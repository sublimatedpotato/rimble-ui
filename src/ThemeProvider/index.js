import React from 'react';
import PropTypes from 'prop-types';

import styled, {
  createGlobalStyle,
  ThemeProvider as StyledThemeProvider,
} from 'styled-components';

import { normalize, transparentize } from 'polished';

import theme from '../theme';

export const Base = styled.div`
  & {
    font-size: 1em;
    font-family: ${props => props.theme.fonts.sansSerif};
    color: ${props => props.theme.colors.copyColor};
  }
`;

const GlobalStyle = createGlobalStyle`
  ${normalize()}
  html {
    box-sizing: border-box;
  }
  *, *:before, *:after {
    box-sizing: inherit;
  }
  ::selection {
    background: ${props => transparentize(0.7, props.theme.colors.primary)};
  }
`;

const ThemeProvider = ({ ...props }) => {
  return (
    <StyledThemeProvider theme={theme} {...props}>
      <React.Fragment>
        <GlobalStyle />
        <Base {...props} />
      </React.Fragment>
    </StyledThemeProvider>
  );
};

ThemeProvider.propTypes = {};

export default ThemeProvider;
