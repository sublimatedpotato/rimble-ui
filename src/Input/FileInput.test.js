import FileInput from './FileInput';
import ThemeProvider from '../ThemeProvider';
import theme from '../theme';

describe('File Input component sanity', () => {
  it('has name', () => {
    expect(FileInput.displayName).toBe('FileInput');
  });

  it('matches default snapshot', () => {
    const component = render(<FileInput />);
    expect(component).toMatchSnapshot();
  });

  it('matches themed snapshot', () => {
    const component = render(
      <ThemeProvider theme={theme}>
        <FileInput />
      </ThemeProvider>
    );
    expect(component).toMatchSnapshot();
  });
});
