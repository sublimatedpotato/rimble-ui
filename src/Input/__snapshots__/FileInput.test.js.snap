// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Input component sanity matches default snapshot 1`] = `
.c2 {
  color: #fff;
  background-color: #999;
  margin: 0;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 32px;
  padding-right: 32px;
  width: medium;
  height: medium;
  border: none;
  border-radius: 4px;
  border-radius: 4px;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  font-size: inherit;
  font-weight: 600;
  --main-color: #4E3FCE;
  --contrast-color: #fff;
  font-size: 1rem;
  height: 3rem;
  min-width: 3rem;
}

.c2 {
  box-sizing: border-box;
}

.c2 {
  -webkit-font-smoothing: antialiased;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  overflow: hidden;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2:hover {
  cursor: pointer;
}

.c2:disabled {
  opacity: 0.5;
  pointer-events: none;
}

.c2 > svg,
.c2 > span.button-text {
  position: relative;
  z-index: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
}

.c2 > svg {
  display: block;
}

.c2 > svg:first-child {
  margin-left: -0.5rem;
  margin-right: .5rem;
}

.c2 > svg:last-child {
  margin-left: .5rem;
  margin-right: -0.5rem;
}

.c2 > svg:first-child:last-child {
  margin: 0;
}

.c2 {
  color: var(--contrast-color);
  background: #FFF;
  position: relative;
  z-index: 0;
  -webkit-transition: all 0.15s ease;
  transition: all 0.15s ease;
}

.c2:hover {
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.2);
}

.c2:hover::before {
  opacity: 0.9;
}

.c2:active {
  box-shadow: none;
  background: #000;
}

.c2::before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 1;
  height: 100%;
  width: 100%;
  background: var(--main-color);
  pointer-events: none;
}

.c1 {
  width: 0;
  height: 0;
  opacity: 0;
  position: absolute;
  overflow: hidden;
  z-index: -1;
}

.c0 {
  display: inline-block;
  position: relative;
  cursor: pointer;
}

<label
  class="c0"
>
  <input
    class="c1"
    type="file"
  />
  <button
    class="c2"
    color="white"
    font-family="sansSerif"
    font-size="inherit"
    font-weight="3"
  >
    <svg
      class=""
      display="block"
      fill="currentcolor"
      height="24px"
      viewBox="0 0 24 24"
      width="24px"
    >
      <path
        d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
      />
    </svg>
    <span
      class="button-text"
    >
      Choose file…
    </span>
  </button>
</label>
`;

exports[`File Input component sanity matches themed snapshot 1`] = `
.c3 {
  color: #fff;
  background-color: #999;
  margin: 0;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 32px;
  padding-right: 32px;
  width: medium;
  height: medium;
  border: none;
  border-radius: 4px;
  border-radius: 4px;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  font-size: inherit;
  font-weight: 600;
  --main-color: #4E3FCE;
  --contrast-color: #fff;
  font-size: 1rem;
  height: 3rem;
  min-width: 3rem;
}

.c3 {
  box-sizing: border-box;
}

.c3 {
  -webkit-font-smoothing: antialiased;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  overflow: hidden;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3:hover {
  cursor: pointer;
}

.c3:disabled {
  opacity: 0.5;
  pointer-events: none;
}

.c3 > svg,
.c3 > span.button-text {
  position: relative;
  z-index: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
}

.c3 > svg {
  display: block;
}

.c3 > svg:first-child {
  margin-left: -0.5rem;
  margin-right: .5rem;
}

.c3 > svg:last-child {
  margin-left: .5rem;
  margin-right: -0.5rem;
}

.c3 > svg:first-child:last-child {
  margin: 0;
}

.c3 {
  color: var(--contrast-color);
  background: #FFF;
  position: relative;
  z-index: 0;
  -webkit-transition: all 0.15s ease;
  transition: all 0.15s ease;
}

.c3:hover {
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.2);
}

.c3:hover::before {
  opacity: 0.9;
}

.c3:active {
  box-shadow: none;
  background: #000;
}

.c3::before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 1;
  height: 100%;
  width: 100%;
  background: var(--main-color);
  pointer-events: none;
}

.c2 {
  width: 0;
  height: 0;
  opacity: 0;
  position: absolute;
  overflow: hidden;
  z-index: -1;
}

.c1 {
  display: inline-block;
  position: relative;
  cursor: pointer;
}

.c0 {
  font-size: 1em;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  color: #3F3D4B;
}

<div
  class="c0"
>
  <label
    class="c1"
  >
    <input
      class="c2"
      type="file"
    />
    <button
      class="c3"
      color="white"
      font-family="sansSerif"
      font-size="inherit"
      font-weight="3"
    >
      <svg
        class=""
        display="block"
        fill="currentcolor"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
      >
        <path
          d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
        />
      </svg>
      <span
        class="button-text"
      >
        Choose file…
      </span>
    </button>
  </label>
</div>
`;
